File: vidgen/pipeline/check_scene_types.py
"""Cross-check scene-type schema drift between Python and TypeScript.

TYPE_MAP / VALID_SCENE_TYPES in render_manifest_builder.py must agree with
the ManifestScene union in remotion/src/types.ts and the switch (shot.type)
dispatch in remotion/src/TikTokVideo.tsx. remotion/scripts/check_scene_types.mjs
is the source-of-truth extractor on the TS side (parses the real AST, not
regex) and writes remotion/scene_types.generated.json; this module runs that
script, then diffs its output against VALID_SCENE_TYPES.

Run directly: python -m vidgen.pipeline.check_scene_types
"""

from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path

from vidgen.pipeline.render_manifest_builder import DIRECT_SNAKE_CASE_SCENE_TYPES, TYPE_MAP

REPO_ROOT = Path(__file__).resolve().parents[2]
REMOTION_ROOT = REPO_ROOT / "remotion"
GENERATED_JSON = REMOTION_ROOT / "scene_types.generated.json"


class SceneTypeDriftError(RuntimeError):
    pass


def _run_extractor() -> None:
    result = subprocess.run(
        ["node", "scripts/check_scene_types.mjs"],
        cwd=REMOTION_ROOT,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        raise SceneTypeDriftError(
            "remotion/scripts/check_scene_types.mjs found drift between "
            "types.ts and TikTokVideo.tsx:\n"
            f"{result.stdout}{result.stderr}"
        )


def check() -> None:
    """Raise SceneTypeDriftError if Python and TypeScript scene types disagree.

    Compares against the render-time snake_case surface only (TYPE_MAP.values()
    | DIRECT_SNAKE_CASE_SCENE_TYPES) — TYPE_MAP's PascalCase keys are an
    author-time convenience with no TS-side counterpart, so they're excluded.
    """
    _run_extractor()

    generated = json.loads(GENERATED_JSON.read_text())
    ts_types = set(generated["sceneTypes"])
    python_types = set(TYPE_MAP.values()) | DIRECT_SNAKE_CASE_SCENE_TYPES

    missing_from_python = ts_types - python_types
    missing_from_ts = python_types - ts_types

    if missing_from_python or missing_from_ts:
        lines = ["scene type cross-check FAILED between Python and TypeScript:"]
        if missing_from_python:
            lines.append(
                f"  - in remotion (types.ts/TikTokVideo.tsx) but not in "
                f"render_manifest_builder's TYPE_MAP/DIRECT_SNAKE_CASE_SCENE_TYPES: "
                f"{sorted(missing_from_python)}"
            )
        if missing_from_ts:
            lines.append(
                f"  - in render_manifest_builder's TYPE_MAP/DIRECT_SNAKE_CASE_SCENE_TYPES "
                f"but not in remotion (types.ts/TikTokVideo.tsx): {sorted(missing_from_ts)}"
            )
        raise SceneTypeDriftError("\n".join(lines))


def main() -> int:
    try:
        check()
    except SceneTypeDriftError as exc:
        print(str(exc), file=sys.stderr)
        return 1
    python_types = set(TYPE_MAP.values()) | DIRECT_SNAKE_CASE_SCENE_TYPES
    print(f"scene type cross-check OK — {len(python_types)} types agree (Python ↔ TypeScript)")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: remotion/scripts/check_scene_types.mjs
// remotion/scripts/check_scene_types.mjs
//
// Cross-checks the three places a "shot type" string has to agree:
//   1. src/types.ts        — ManifestScene union, source of truth
//   2. src/TikTokVideo.tsx — switch (shot.type) render dispatch
//   3. ../vidgen/pipeline/render_manifest_builder.py — TYPE_MAP values
//
// Parses via the real project/checker (typescript's native TS7 API), not
// regex, so it reads the same AST tsc itself would type-check against.
// Also writes remotion/scene_types.generated.json so Python can diff
// against the same source of truth without re-parsing TypeScript.
//
// Usage: node scripts/check_scene_types.mjs [--check-only]
//   --check-only   skip writing scene_types.generated.json; only exit
//                  non-zero on drift (used by `tsc --noEmit`-style CI steps
//                  that shouldn't touch the working tree)

import { readFileSync, writeFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { createRequire } from "node:module";

const require = createRequire(import.meta.url);
const { API } = require("typescript/unstable/sync");
const is = require("typescript/unstable/ast/is");

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const REMOTION_ROOT = path.resolve(__dirname, "..");
const OUTPUT_PATH = path.join(REMOTION_ROOT, "scene_types.generated.json");

// Walk ManifestScene's union members and collect each `type: "literal"` string.
function extractManifestSceneTypes(sourceFile) {
  let unionNode = null;
  sourceFile.forEachChild((node) => {
    if (unionNode) return;
    if (is.isTypeAliasDeclaration(node) && node.name.text === "ManifestScene") {
      unionNode = node;
    }
  });
  if (!unionNode) {
    throw new Error("Could not find `type ManifestScene = ...` — check src/types.ts");
  }
  if (!is.isUnionTypeNode(unionNode.type)) {
    throw new Error("ManifestScene is not a union type — check src/types.ts");
  }

  const types = [];
  for (const member of unionNode.type.types) {
    if (!is.isTypeLiteralNode(member)) continue;
    member.forEachChild((prop) => {
      if (
        is.isPropertySignatureDeclaration(prop) &&
        prop.name &&
        is.isIdentifier(prop.name) &&
        prop.name.text === "type" &&
        prop.type &&
        is.isLiteralTypeNode(prop.type) &&
        is.isStringLiteral(prop.type.literal)
      ) {
        types.push(prop.type.literal.text);
      }
    });
  }
  if (types.length === 0) {
    throw new Error('Found no `type: "..."` members on ManifestScene — check src/types.ts');
  }
  return types;
}

// Walk the `switch (shot.type)` in TikTokVideo.tsx and collect each case's string literal.
function extractSwitchCaseTypes(sourceFile) {
  let switchNode = null;
  function visit(node) {
    if (switchNode) return;
    if (
      is.isSwitchStatement(node) &&
      is.isPropertyAccessExpression(node.expression) &&
      node.expression.name.text === "type"
    ) {
      switchNode = node;
      return;
    }
    node.forEachChild(visit);
  }
  visit(sourceFile);
  if (!switchNode) {
    throw new Error("Found no `switch (shot.type)` statement — check src/TikTokVideo.tsx");
  }

  return switchNode.caseBlock.clauses
    .filter(is.isCaseClause)
    .map((clause) => clause.expression)
    .filter(is.isStringLiteral)
    .map((lit) => lit.text);
}

function main() {
  const checkOnly = process.argv.includes("--check-only");

  const api = new API({ cwd: REMOTION_ROOT });
  let unionTypes, switchTypes;
  try {
    const snapshot = api.updateSnapshot({ openProjects: ["tsconfig.json"] });
    const project = snapshot.getProject("tsconfig.json");
    if (!project) {
      throw new Error("Could not load remotion/tsconfig.json as a project");
    }

    const typesSourceFile = project.program.getSourceFile("src/types.ts");
    const videoSourceFile = project.program.getSourceFile("src/TikTokVideo.tsx");
    if (!typesSourceFile) throw new Error("Could not load src/types.ts from the program");
    if (!videoSourceFile) throw new Error("Could not load src/TikTokVideo.tsx from the program");

    unionTypes = extractManifestSceneTypes(typesSourceFile);
    switchTypes = extractSwitchCaseTypes(videoSourceFile);
  } finally {
    api.close();
  }

  const unionSet = new Set(unionTypes);
  const switchSet = new Set(switchTypes);

  const errors = [];

  const dupUnion = unionTypes.filter((t, i) => unionTypes.indexOf(t) !== i);
  if (dupUnion.length > 0) {
    errors.push(`Duplicate type literals in ManifestScene union: ${[...new Set(dupUnion)].join(", ")}`);
  }

  const missingFromSwitch = [...unionSet].filter((t) => !switchSet.has(t));
  if (missingFromSwitch.length > 0) {
    errors.push(
      `Types in ManifestScene union but not handled by the TikTokVideo.tsx switch: ${missingFromSwitch.join(", ")}`
    );
  }

  const missingFromUnion = [...switchSet].filter((t) => !unionSet.has(t));
  if (missingFromUnion.length > 0) {
    errors.push(
      `Types handled by the TikTokVideo.tsx switch but absent from the ManifestScene union: ${missingFromUnion.join(", ")}`
    );
  }

  if (errors.length > 0) {
    console.error("scene type cross-check FAILED:\n");
    for (const e of errors) console.error(`  - ${e}`);
    console.error(`\nManifestScene union (${unionTypes.length}): src/types.ts`);
    console.error(`TikTokVideo.tsx switch (${switchTypes.length}): src/TikTokVideo.tsx`);
    process.exit(1);
  }

  console.log(
    `scene type cross-check OK — ${unionTypes.length} scene types agree between types.ts and TikTokVideo.tsx`
  );

  if (!checkOnly) {
    const payload = {
      // Sorted for a stable diff; this file is generated, do not hand-edit.
      sceneTypes: [...unionSet].sort(),
      generatedFrom: ["src/types.ts:ManifestScene", "src/TikTokVideo.tsx:switch(shot.type)"],
    };
    writeFileSync(OUTPUT_PATH, JSON.stringify(payload, null, 2) + "\n");
    console.log(`wrote ${path.relative(REMOTION_ROOT, OUTPUT_PATH)}`);
  }
}

main();
